cmake/cmaize/project/project.cmake

```diff
@@ -0,0 +1,145 @@
+    #]]
+    cpp_attr(CMaizeProject languages)
+
+    #[[[
+    # :type: list[PackageManager]
+    #
+    # Package managers used by the project.
+    #]]
+    cpp_attr(CMaizeProject package_managers)
+
+    #[[[
+    # :type: list[Target]
+    #
+    # Targets that are a part of the project.
+    #]]
+    cpp_attr(CMaizeProject targets)
+
+    #[[[
+    # Creates a ``CMaizeProject`` object and underlying CMake project, if
+    # necessary.
+    #
+    # The underlying CMake project is only created if the given project name
+    # does not match the CMake project name in ``PROJECT_NAME``.
+    #
+    # :param self: The constructed object.
+    # :type self: CMaizeProject
+    # :param _ctor_name: Name of the project.
+    # :type _ctor_name: desc
+    #
+    # :Keyword Arguments:
+    #    * **VERSION** (*desc*) --
+    #      Version string of non-negative integers of the form
+    #      ``<major>[.<minor>[.<patch>[.<tweak>] ] ]``. Populates the
+    #      version attributes of the ``ProjectSpecification``. Parallels
+    #      the ``VERSION <version>`` keyword for CMake ``project()``
+    #      (`link <cmake_project_>`__), and the value is passed to
+    #      a CMake ``project()`` call if the project does not exist yet.
+    #    * **DESCRIPTION** (*desc*) --
+    #      Description of the project. Parallels the 
+    #      ``DESCRIPTION <project-description-string>`` keyword for 
+    #      CMake ``project()`` (`link <cmake_project_>`__), and the value
+    #      is passed to a CMake ``project()`` call if the project does not
+    #      exist yet.
+    #    * **HOMEPAGE_URL** (*desc*) --
+    #      Homepage URL for the project. Parallels the
+    #      ``HOMEPAGE_URL <url-string>`` keyword for CMake ``project()``
+    #      (`link <cmake_project_>`__), and the value is passed to
+    #      a CMake ``project()`` call if the project does not exist yet.
+    #    * **LANGUAGES** (*list[desc]*) --
+    #      Languages supported by the project. These languages are passed to
+    #      the LANGUAGES keyword of the CMake ``project()`` call if the project
+    #      does not exist yet.
+    #
+    # :returns: ``self`` will be set to the newly constructed
+    #           ``CMaizeProject`` object.
+    # :rtype: CMaizeProject
+    # 
+    # .. Reference definitions
+    # .. _cmake_project: https://cmake.org/cmake/help/latest/command/project.html
+    #]]
+    cpp_constructor(CTOR CMaizeProject str args)
+    function("${CTOR}" self _ctor_name)
+
+        set(_ctor_lists LANGUAGES)
+        cmake_parse_arguments(_ctor "" "" "${_ctor_lists}" ${ARGN})
+
+        CMaizeProject(SET "${self}" name "${_ctor_name}")
+
+        CMaizeProject(SET "${self}" languages "${_ctor_LANGUAGES}")
+
+        # If the project name is different than the current project, assume
+        # a new CMake project needs to be created
+        if(NOT "${_ctor_name}" STREQUAL "${PROJECT_NAME}")
+            project("${_ctor_name}" ${ARGN})
+        endif()
```

Comment:
Maybe throw an error instead?

Reply:
Ah, I didn't realize calling `project` was going to be required. This can definitely be an error instead.